metricbeat/module/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,127 @@
+package elasticsearch
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	"github.com/elastic/beats/metricbeat/helper"
+)
+
+// Global clusterIdCache. Assumption is that the same node id never can belong to a different cluster id
+var clusterIdCache = map[string]string{}
+
+type infoStruct struct {
+	ClusterName string `json:"cluster_name"`
+	ClusterId   string `json:"cluster_uuid"`
+}
+
+func GetClusterId(http *helper.HTTP, uri string, nodeId string) (string, error) {
+
+	// Check if cluster id already cached. If yes, return it.
+	if clusterId, ok := clusterIdCache[nodeId]; ok {
+		return clusterId, nil
+	}
+
+	// Makes sure the http uri is reset to its inital value
+	defer http.SetURI(uri)
+
+	info, err := getInfo(http, uri)
+	if err != nil {
+		return "", err
+	}
+
+	clusterIdCache[nodeId] = info.ClusterId
+	return info.ClusterId, nil
+}
+
+// IsMaster checks if the given node host is a master node
+//
+// The detection of the master is done in two steps:
+// * Fetch node name from /_nodes/_local/name
+// * Fetch current master name from cluster state /_cluster/state/master_node
+//
+// The two names are compared
+func IsMaster(http *helper.HTTP, uri string) (bool, error) {
+	// Makes sure the http uri is reset to its inital value
+	defer http.SetURI(uri)
+
+	node, err := getNodeName(http, uri)
+	if err != nil {
+		return false, err
+	}
+
+	master, err := getMasterName(http, uri)
+	if err != nil {
+		return false, err
+	}
+
+	return master == node, nil
+}
+
+func getNodeName(http *helper.HTTP, uri string) (string, error) {
+
+	// Parses the uri to replace the path
+	u, _ := url.Parse(uri)
+	u.Path = "/_nodes/_local/nodes"
+
+	// Http helper includes the HostData with username and password
+	http.SetURI(u.String())
+	content, err := http.FetchContent()
+	if err != nil {
+		return "", err
+	}
+
+	nodesStruct := struct {
+		Nodes map[string]interface{} `json:"nodes"`
+	}{}
+
+	json.Unmarshal(content, &nodesStruct)
+
+	// _local will only fetch one node info. First entry is node name
+	for k, _ := range nodesStruct.Nodes {
+		return k, nil
+	}
+	return "", fmt.Errorf("No local node found")
+}
+
+func getMasterName(http *helper.HTTP, uri string) (string, error) {
+
+	// Parses the uri to replace the path
+	u, _ := url.Parse(uri)
+	u.Path = "_cluster/state/master_node"
```

Comment:
We can probably call `GET /_cluster/state/master_node?local=true`  to avoid having to bounce to the elected master to find out what this node thinks. Since we have to asynchronously fetch this, it's a race condition regardless. If it's wrong, it'll fix itself on the next pass.

Reply:
Nice, wasn't aware of this param. That means we also get an answer in case the node is disconnected I assume.